### PR TITLE
Tweaks to improve CI performance

### PR DESCRIPTION
**Pull request by rcoh**
- Run Pytest in parallel
- Only run `--qa` for the develop branch
- Public repo xfails actually fail if they pass.

### CODE REVIEW

**Comment by brendongo** [comment]
I think instead of only running --qa in develop we only run the full range of --qa in develop but it might still make some sense to run some of the qa tests on every PR?

**Comment by mschwager** [comment]
> I think instead of only running --qa in develop we only run the full range of --qa in develop but it might still make some sense to run some of the qa tests on every PR?

Another option may be to run a random 20% of the QA tests every time. That way any given commit/PR will be fast, and we'll be likely to keep the same coverage over time.

**Comment by rcoh** [comment]
I don't really love any sort of non-determinism in tests. We'd have to make it based on branch name or something. I think perfect is the enemy of an improvement here.

File: .github/workflows/bento.yml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+jobs:
```
Comment (brendongo) [inline .github/workflows/bento.yml]: Lint workflow already runs bento

File: semgrep/Makefile [hunk 2]
```diff
@@ -14,6 +14,17 @@ install:
 qa-test:
 	pipenv run pytest -v --tb=short tests/ --qa
 
+
+
+.PHONY: ci-test
```
Comment (brendongo) [inline semgrep/Makefile:19]: Can we have the switch happen in the github workflow instead of makefile.

File: semgrep/tox.ini [hunk 3]
```diff
@@ -5,7 +5,7 @@ envlist = py36, py37, py38
 [testenv]
 commands =
     pipenv install --sequential --dev
-    pipenv run pytest tests --qa
+    pipenv run pytest -n auto
```
Comment (brendongo) [inline semgrep/tox.ini:8]: what does auto do?

Comment (rcoh) [reply semgrep/tox.ini:8]: num cores

**Comment by brendongo** [comment]
If we want determinism then we can have a a select subset of the public_repos test to run. But we should definitely not be in the world where none of the qa-tests run in a PR. The core issue was not that we were running qa-tests on every PR but the fact that there are > 150 repos being tests in every PR. I think cutting it down to ~20 and having the rest run on develop would be fine. We also want to keep running the test_semgrep_rules_repo.py running on every PR.

**Comment by brendongo** [comment]
I would even be comfortable with having all public repo tests be only run on develop but I think the test_semgrep_rules_repo should be PR blocking. If informs us a lot of regressions whenever pfff submodule is updated and that happens quite often.

**Comment by rcoh** [comment]
Ok! I can move that into mainline tests. Re: GitHub actions I considered that but, conditionals there are really janky & impossible to share across workflows. This way, there is one command to rule them all.

**Comment by brendongo** [comment]
> Re: GitHub actions I considered that but, conditionals there are really janky & impossible to share across workflows. This way, there is one command to rule them all.

Fair. Makefile it is

File: semgrep/tests/conftest.py [hunk 4]
```diff
@@ -220,3 +221,12 @@ def pytest_configure(config):
 def pytest_runtest_setup(item):
     if item.get_closest_marker("qa") and not item.config.getoption("--qa"):
         pytest.skip("skipping QA tests, add --qa flag to run them")
+
+
+# Add `public_repo_url` as a parameter to your test to use this. It generates 1 test case per repo url in `ALL_REPOS`
+def pytest_generate_tests(metafunc):
```
Comment (brendongo) [inline semgrep/tests/conftest.py:227]: Can we simplify this to just have two different funtions: one that runs the 5 repos and one that runs a lot of repos. Not sure why this needs to be a fixture and have a separate file with all the repos? Seems like a lot of complexity.

Comment (brendongo) [reply semgrep/tests/conftest.py:227]: Discussed offline. We can revisit re: complexity when it becomes a problem merging as is.

**Review by brendongo — APPROVED** [review]

**Pull request merged**